### PR TITLE
Fix show gpus when user does not have AWS credentials

### LINKED ISSUE #1710: `sky show-gpus` should not assume AWS credentials to fetch AZ mappings
On a fresh docker install
```
pip install skypilot[lambda]
sky show-gpus
```
shows
```
...
I 02-21 20:24:39 aws_catalog.py:53] Fetching availability zones mapping for AWS...
--- Logging error ---
....
  File "/opt/conda/lib/python3.9/site-packages/sky/clouds/service_catalog/data_fetchers/fetch_aws.py", line 69, in get_enabled_regions
    user_cloud_regions = aws_client.describe_regions()['Regions']
  File "/opt/conda/lib/python3.9/site-packages/botocore/client.py", line 530, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/opt/conda/lib/python3.9/site-packages/botocore/client.py", line 943, in _make_api_call
    http, parsed_response = self._make_request(
  File "/opt/conda/lib/python3.9/site-packages/botocore/client.py", line 966, in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
  File "/opt/conda/lib/python3.9/site-packages/botocore/endpoint.py", line 119, in make_request
    return self._send_request(request_dict, operation_model)
  File "/opt/conda/lib/python3.9/site-packages/botocore/endpoint.py", line 198, in _send_request
    request = self.create_request(request_dict, operation_model)
  File "/opt/conda/lib/python3.9/site-packages/botocore/endpoint.py", line 134, in create_request
    self._event_emitter.emit(
  File "/opt/conda/lib/python3.9/site-packages/botocore/hooks.py", line 412, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/botocore/hooks.py", line 256, in emit
    return self._emit(event_name, kwargs)
  File "/opt/conda/lib/python3.9/site-packages/botocore/hooks.py", line 239, in _emit
    response = handler(**kwargs)
  File "/opt/conda/lib/python3.9/site-packages/botocore/signers.py", line 105, in handler
    return self.sign(operation_name, request)
  File "/opt/conda/lib/python3.9/site-packages/botocore/signers.py", line 189, in sign
    auth.add_auth(request)
  File "/opt/conda/lib/python3.9/site-packages/botocore/auth.py", line 418, in add_auth
    raise NoCredentialsError()
botocore.exceptions.NoCredentialsError: Unable to locate credentials
```
We should perhaps check other call paths that will by default fetch AWS AZ mappings.

cc @WoosukKwon @Michaelvll

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #1710 

Tested (run the relevant ones):

Without credential
- [x] `sky show-gpus` (1.4 seconds; origin: 1.4 seconds)
- [x] `sky show-gpus --all`
- [x] `sky show-gpus --cloud aws --region us-east-1` (1.4 seconds; origin: 1.4 seconds)

With credential
- [x] `sky show-gpus` (1.4 seconds; origin: 1.4 seconds)
- [x] `sky show-gpus --all`
- [x] `sky show-gpus --cloud aws --region us-east-1` (1.2 seconds; origin: 1.2 seconds)

Tests:
- [x] Launch an AWS cluster; remove the AWS credentials; `sky status` and `sky status -r`
- [x] `pytest tests/test_smoke.py`
- [x] `pytest tests/test_smoke.py` with no AWS credentials
- [x] `pytest tests/test_smoke.py --aws`

### CODE REVIEW

File: sky/clouds/service_catalog/aws_catalog.py [hunk 1]
```diff
@@ -61,25 +70,52 @@ def _apply_az_mapping(df: 'pd.DataFrame') -> 'pd.DataFrame':
...
+    def wrapper(*args, **kwargs):
+        with _apply_az_mapping_lock:
+            _apply_az_mapping()
+        return func(*args, **kwargs)
 
-_df = _apply_az_mapping(_df)
+    return wrapper
 
 
+@_apply_az_mapping_decorator
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:90]: Does this func need the mapping?

File: sky/clouds/service_catalog/aws_catalog.py [hunk 2]
```diff
@@ -61,25 +70,52 @@ def _apply_az_mapping(df: 'pd.DataFrame') -> 'pd.DataFrame':
...
 def instance_type_exists(instance_type: str) -> bool:
     return common.instance_type_exists_impl(_df, instance_type)
 
 
+# Helper functions for `validate_region_zone`.
+def _validate_region(
+        region: Optional[str]) -> Tuple[Optional[str], Optional[str]]:
+    # When zone is not specified, we don't need to apply the AZ mapping.
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:98]:
```suggestion
    # When zone is not specified, we don't need to apply the AZ mapping (note that no decorator is applied to this func).
```

File: sky/clouds/service_catalog/aws_catalog.py [hunk 3]
```diff
@@ -61,25 +70,52 @@ def _apply_az_mapping(df: 'pd.DataFrame') -> 'pd.DataFrame':
...
+    # This is useful for `sky show-gpus` where user may not have access to
+    # AWS.
+    return common.validate_region_zone_impl('aws', _df, region, None)
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:101]:
```suggestion
    return common.validate_region_zone_impl('aws', _df, region, zone=None)
```

File: sky/clouds/service_catalog/aws_catalog.py [hunk 4]
```diff
@@ -61,25 +70,52 @@ def _apply_az_mapping(df: 'pd.DataFrame') -> 'pd.DataFrame':
...
-_df = _apply_az_mapping(_df)
+    return wrapper
 
 
+@_apply_az_mapping_decorator
 def instance_type_exists(instance_type: str) -> bool:
     return common.instance_type_exists_impl(_df, instance_type)
 
 
+# Helper functions for `validate_region_zone`.
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:95]: Minor preference: I feel like it's clearer to keep these internal to validate_region_zone() (shallow methods). Up to you though.

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll! Some comments.

**Comment by Michaelvll** [comment]
Sorry for changing the implementation @concretevitamin. I found that the previous implementation will make the `validate_region_zone` inaccurate for the user with AWS credentials, as we may not exclude the regions not supported by the user.

File: sky/clouds/service_catalog/aws_catalog.py [hunk 5]
```diff
@@ -35,17 +36,28 @@
...
 
 
-def _apply_az_mapping(df: 'pd.DataFrame') -> 'pd.DataFrame':
+def _apply_az_mapping(df: pd.DataFrame) -> pd.DataFrame:
     """Maps zone IDs (use1-az1) to zone names (us-east-1x).
 
-    Such mappings are account-specific and determined by AWS.
+    Such mappings are account-specific and determined by AWS. The mpaaing will
+    also remove the regions that are not supported by the user account. If the
+    user does not have AWS credentials configured, we use a default mapping.
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:44]:
```suggestion
    Such mappings are account-specific and determined by AWS. We fetch the mappings from AWS, which requires AWS credentials. If the user does not have AWS credentials configured, we use a default mapping.
    
    The mappings will also serve to remove from '_df' the regions that are not supported by the user account. 
```
Actually, can we briefly explain what "default mapping" means, and why that is ok?

File: sky/clouds/service_catalog/aws_catalog.py [hunk 6]
```diff
@@ -35,17 +36,28 @@
...
         with the zone name (e.g. us-east-1a).
     """
     az_mapping_path = common.get_catalog_path('aws/az_mappings.csv')
     if not os.path.exists(az_mapping_path):
+        aws_enabled, _ = aws.AWS().check_credentials()
+        if not aws_enabled:
+            logger.debug(
+                f'{colorama.Style.DIM}Availability zone mapping is not '
+                f'available because AWS credentials are not configured. '
+                f'Using the default mapping.{colorama.Style.RESET_ALL}')
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:57]: There may be an edge case where `check_credentials()` returns False, but in fact `fetch_aws.fetch_availability_zone_mappings()` has the necessary credential files and go through? E.g., the `aws` CLI is not installed, but https://github.com/skypilot-org/skypilot/blob/d9fd2794c2db1befa529fd7e1b7e5aae7b23d1eb/sky/clouds/service_catalog/data_fetchers/fetch_aws.py#L106 may go through.

How about we try-catch L69?

File: sky/clouds/service_catalog/aws_catalog.py (hunk 5 above)
Comment (Michaelvll) [reply sky/clouds/service_catalog/aws_catalog.py]: Modified and added the reason for default mapping. Thanks!

File: sky/clouds/service_catalog/aws_catalog.py [hunk 7]
```diff
@@ -35,26 +37,52 @@
...
+    not supported by the user account.
 
     Returns:
         A dataframe with column 'AvailabilityZone' that's correctly replaced
         with the zone name (e.g. us-east-1a).
     """
-    az_mapping_path = common.get_catalog_path('aws/az_mappings.csv')
+    aws_enabled, _ = aws.AWS().check_credentials()
+    if aws_enabled:
+        user_identity = aws.AWS().get_current_user_identity()
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:59]: This seems visibly slower on my laptop: 0.9s -> 2.1s for `time sky show-gpus --cloud aws`.

File: sky/clouds/service_catalog/aws_catalog.py [hunk 8]
```diff
@@ -35,26 +37,52 @@
...
+                ray.init()
+            az_mappings = fetch_aws.fetch_availability_zone_mappings()
+        else:
+            logger.debug('Failed to fetch availability zone mappings (using '
+                         'default mapping)')
+            dummy_az_name = _df.rename(
+                columns={'AvailabilityZone': 'AvailabilityZoneName'
+                        })['AvailabilityZoneName']
+            az_mappings = pd.concat([_df['AvailabilityZone'], dummy_az_name],
+                                    axis=1)
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/clouds/service_catalog/aws_catalog.py]: We are now returning the df directly, as the two expressions here take a lot of time.

File: sky/clouds/service_catalog/aws_catalog.py (hunk 7 above)
Comment (Michaelvll) [reply sky/clouds/service_catalog/aws_catalog.py]: Just optimized it for `show-gpus`. Please feel free to have another try. Thanks!

File: sky/clouds/service_catalog/config.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+import contextlib
+import threading
+
+from sky.utils import common_utils
+
+local_config = threading.local()
+# Whether the caller requires the catalog to be narrowed down
+# to the account-specific catalog or just the raw catalog
+# fetched from SkyPilot catalog service.
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/config.py:11]:
```suggestion
# Whether the caller requires the catalog to be narrowed down
# to the account-specific catalog (e.g., removing regions not enabled for the current account) or just the raw catalog
# fetched from SkyPilot catalog service. The former is used for launching clusters, while the latter for commands like `show-gpus`.
```
Nice!
```

File: sky/clouds/service_catalog/config.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    finally:
+        local_config.use_default_catalog = old_value
+
+
+def get_use_default_catalog() -> bool:
+    return local_config.use_default_catalog
+
+
+# Decorator to disable account-specific catalog.
+def use_default_catalog(func):
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/config.py:31]:
```suggestion
def use_default_catalog(func):
    """Decorator: disable fetching account-specific catalog which need credentials."""
```

File: sky/clouds/service_catalog/config.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+local_config = threading.local()
+# Whether the caller requires the catalog to be narrowed down
+# to the account-specific catalog or just the raw catalog
+# fetched from SkyPilot catalog service.
+local_config.use_default_catalog = False
+
+
+@contextlib.contextmanager
+def set_use_default_catalog(name: str, to_use: bool):
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/config.py:16]: - to_use -> value?
- why do we have a `name` arg?
- seems like we can prefix this func with `_`

File: sky/clouds/service_catalog/config.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        local_config.use_default_catalog = old_value
+
+
+def get_use_default_catalog() -> bool:
+    return local_config.use_default_catalog
+
+
+# Decorator to disable account-specific catalog.
+def use_default_catalog(func):
+    return common_utils.make_decorator(set_use_default_catalog,
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/config.py:32]: First time reading this func. What does it mean? Make `func` a decorator and run `set_use_default_catalog` beforehand..?

File: sky/clouds/service_catalog/__init__.py [hunk 13]
```diff
@@ -309,6 +314,8 @@ def is_image_tag_valid(tag: str,
     # Images
     'get_image_id_from_tag',
     'is_image_tag_valid',
+    # Configuring
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/__init__.py:317]:
```suggestion
    # Configuration.
```

File: sky/clouds/service_catalog/aws_catalog.py [hunk 14]
```diff
@@ -31,58 +35,79 @@
...
     not supported by the user account.
 
     Returns:
         A dataframe with column 'AvailabilityZone' that's correctly replaced
         with the zone name (e.g. us-east-1a).
     """
-    aws_enabled, _ = aws.AWS().check_credentials()
-    if aws_enabled:
-        user_identity = aws.AWS().get_current_user_identity()
+    # The caller should guarantee that the aws credentials are valid.
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:67]: L67 and L50 seem to contradict L72?

File: sky/clouds/service_catalog/aws_catalog.py [hunk 15]
```diff
@@ -31,58 +35,79 @@
...
+_default_df = common.read_catalog('aws/vms.csv',
+                                  pull_frequency_hours=_PULL_FREQUENCY_HOURS)
+_user_df = None
+_apply_az_mapping_lock = threading.Lock()
+
 _image_df = common.read_catalog('aws/images.csv',
                                 pull_frequency_hours=_PULL_FREQUENCY_HOURS)
 
 
 def _apply_az_mapping(df: pd.DataFrame) -> pd.DataFrame:
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:47]: `_fetch_and_apply_az_mapping`?

File: sky/clouds/service_catalog/aws_catalog.py [hunk 16]
```diff
@@ -31,58 +35,79 @@
...
+    except exceptions.CloudUserIdentityError:
+        glob_name = common.get_catalog_path('aws/az_mappings-*.csv')
+        # Find the most recent file that matches the glob.
+        glob_files = glob.glob(glob_name)
+        if glob_files:
+            glob_files.sort(key=os.path.getmtime)
+            aws_user_hash = os.path.basename(glob_files[-1]).split('-')[1]
+            aws_user_hash = aws_user_hash.split('.')[0]
+        else:
+            aws_user_hash = 'default'
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:81]: Curious why we find the latest file, vs. just use `default`? Document the reason?

File: sky/clouds/service_catalog/aws_catalog.py [hunk 17]
```diff
@@ -31,58 +35,79 @@
 # skypilot-catalog/.github/workflows/update-aws-catalog.yml
 _PULL_FREQUENCY_HOURS = 7
 
-_df = common.read_catalog('aws/vms.csv',
-                          pull_frequency_hours=_PULL_FREQUENCY_HOURS)
+_default_df = common.read_catalog('aws/vms.csv',
+                                  pull_frequency_hours=_PULL_FREQUENCY_HOURS)
+_user_df = None
+_apply_az_mapping_lock = threading.Lock()
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:41]:
```suggestion
# The main catalog dataframe.
#   - _default_df: default non-account-specific catalog
#   - _user_df: account-specific catalog (i.e., regions that the account doesn't have enabled are dropped; AZ mapping is applied, etc.) Creating this requires AWS credentials. It is created at most once (and cached) per a process' lifetime.
# '_apply_az_mapping_lock' protects reading/writing '_user_df'.
_default_df = common.read_catalog('aws/vms.csv',
                                  pull_frequency_hours=_PULL_FREQUENCY_HOURS)
_user_df = None
_apply_az_mapping_lock = threading.Lock()
```

File: sky/clouds/service_catalog/aws_catalog.py [hunk 18]
```diff
@@ -31,58 +35,79 @@
...
-    else:
-        aws_user_hash = 'default'
+    except exceptions.CloudUserIdentityError:
+        glob_name = common.get_catalog_path('aws/az_mappings-*.csv')
+        # Find the most recent file that matches the glob.
+        glob_files = glob.glob(glob_name)
+        if glob_files:
+            glob_files.sort(key=os.path.getmtime)
+            aws_user_hash = os.path.basename(glob_files[-1]).split('-')[1]
+            aws_user_hash = aws_user_hash.split('.')[0]
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:79]: document that this may return `default`

File: sky/clouds/service_catalog/aws_catalog.py [hunk 19]
```diff
@@ -31,58 +35,79 @@
...
-        else:
-            logger.debug('Failed to fetch availability zone mappings (using '
-                         'default mapping)')
-            dummy_az_name = _df.rename(
-                columns={'AvailabilityZone': 'AvailabilityZoneName'
-                        })['AvailabilityZoneName']
-            az_mappings = pd.concat([_df['AvailabilityZone'], dummy_az_name],
-                                    axis=1)
+        az_mappings = None
+        if aws_user_hash != 'default':
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:90]:
```suggestion
        if aws_user_hash != 'default':  # short-circuit check_credentials()
```

File: sky/clouds/service_catalog/aws_catalog.py [hunk 20]
```diff
@@ -31,58 +35,79 @@
...
+            aws_enabled, _ = aws.AWS.check_credentials()
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:91]: Shall we skip `check_credentials` which is more stringent than the permissions needed to fetch the mappings? May also reduce number of nested `if`.

File: sky/clouds/service_catalog/aws_catalog.py [hunk 21]
```diff
@@ -31,58 +35,79 @@
...
+                logger.info(f'{colorama.Style.DIM}Fetching availability zones '
+                            f'mapping for AWS...{colorama.Style.RESET_ALL}')
+                with ux_utils.suppress_output():
+                    ray.init()
+                az_mappings = fetch_aws.fetch_availability_zone_mappings()
+
+        if az_mappings is None:
+            # Returning the original dataframe directly, as no az mapping
+            # is available. The caller should handle the case where the
+            # credentials are invalid.
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:106]:
```suggestion
        else:
            # Returning the original dataframe directly, as no cloud identity can be fetched which suggests there are no credentials.
```

File: sky/clouds/service_catalog/aws_catalog.py [hunk 22]
```diff
@@ -31,58 +35,79 @@
...
+            if 'AvailabilityZoneName' in df.columns:
+                df = df.drop(columns=['AvailabilityZone']).rename(
+                    columns={'AvailabilityZoneName': 'AvailabilityZone'})
```
Comment (concretevitamin) [inline sky/clouds/service_catalog/aws_catalog.py:109]: Confusing?

File: tests/test_optimizer_dryruns.py [hunk 23]
```diff
@@ -47,6 +47,9 @@ def _make_resources(
         lambda: enabled_clouds,
     )
     monkeypatch.setattr('sky.check.check', lambda *_args, **_kwargs: None)
+    monkeypatch.setattr(
+        'sky.clouds.service_catalog.config.get_use_default_catalog',
+        lambda: True)
```
Comment (concretevitamin) [inline tests/test_optimizer_dryruns.py:52]: ?

File: sky/utils/db_utils.py [hunk 24]
```diff
@@ -16,8 +16,13 @@ def add_column_to_table(
         if row[1] == column_name:
             break
     else:
-        cursor.execute(f'ALTER TABLE {table_name} '
-                       f'ADD COLUMN {column_name} {column_type}')
+        try:
+            cursor.execute(f'ALTER TABLE {table_name} '
+                           f'ADD COLUMN {column_name} {column_type}')
+        except sqlite3.OperationalError:
```
Comment (concretevitamin) [inline sky/utils/db_utils.py:22]: Can we additionally check it's due to duplicating the column addition?

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll. Some comments. May want to run some smoke tests.

File: sky/clouds/aws.py [hunk 25]
```diff
@@ -463,6 +463,7 @@ def check_credentials(cls) -> Tuple[bool, Optional[str]]:
 
         # Fetch the AWS availability zones mapping from ID to name.
         from sky.clouds.service_catalog import aws_catalog  # pylint: disable=import-outside-toplevel,unused-import
+        aws_catalog.get_default_instance_type()
```
Comment (concretevitamin) [inline sky/clouds/aws.py:466]: Does this func trigger fetching? May want to comment.

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks @Michaelvll.

**Pull request merged**